src/main.cpp

```diff
@@ -90,87 +91,197 @@ namespace
-            //   ]
-            // }
-            for (auto&& prefix : config->at("prefixes")) {
-                // If the metadata attribute starts with the prefix, then verify that the user
-                // can modify the metadata attribute.
-                if (boost::starts_with(input->arg3, prefix.get_ref<const std::string&>())) {
-                    // The "admin_only" flag supersedes the "editors" configuration option.
-                    if (config->count("admin_only") && config->at("admin_only").get<bool>()) {
-                        return user_is_administrator(*rei.rsComm);
-                    }
-
-                    namespace adm = irods::experimental::administration;
-
-                    const adm::user user{rei.uoic->userName, rei.uoic->rodsZone};
-
-                    for (auto&& editor : config->at("editors")) {
-                        if (const auto& type = editor.at("type").get_ref<const std::string&>(); type == "group") {
-                            const adm::group group{editor.at("name").get_ref<const std::string&>()};
-
-                            if (adm::server::user_is_member_of_group(*rei.rsComm, group, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                        else if (type == "user") {
-                            if (editor.at("name").get_ref<const std::string&>() == adm::server::local_unique_name(*rei.rsComm, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                    }
-
-                    // At this point, the user is not an administrator and they aren't a member of
-                    // the editors list. Therefore, we return an error because the user is attempting to
-                    // modify metadata in a guarded namespace.
-                    return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
-                }
-            }
-        }
-        catch (const json::exception&) {
-            // clang-format off
+		_exists = (_rule_name == "pep_api_mod_avu_metadata_pre") ||
+		          (_rule_name == "pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto list_rules(irods::default_re_ctx&, std::vector<std::string>& _rules) -> irods::error
+	{
+		_rules.push_back("pep_api_mod_avu_metadata_pre");
+		_rules.push_back("pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto metadata_guard_handling(const std::string& _object_name, ruleExecInfo_t& rei) -> irods::error
+	{
+		const auto config = load_plugin_config(rei);
+		const auto& prefixes = config->at("prefixes");
+		const auto prefix_matched =
+			std::any_of(prefixes.cbegin(), prefixes.cend(), [&_object_name](const json& prefix) {
+				return boost::starts_with(_object_name, prefix.get_ref<const std::string&>());
+			});
+		if (prefix_matched) {
```

Comment:
Could lose a level of indentation here if the logic is inverted and we return early.